lib/package_config_files.c

```diff
@@ -170,15 +170,30 @@ xbps_entry_install_conf_file(struct xbps_handle *xhp,
 		/*
 		 * Orig = X, Curr = X, New = Y
 		 *
-		 * Install new file (installed file hasn't been modified).
+		 * Install new file (installed file hasn't been modified) if
+		 * configuration option keepconfig is NOT set.
+		 *
+		 * Install new file as <file>.new-<version> if configuration
+		 * option perserveconfig is set.
 		 */
 		} else if ((strcmp(sha256_orig, sha256_cur) == 0) &&
 			   (strcmp(sha256_orig, sha256_new)) &&
 			   (strcmp(sha256_cur, sha256_new))) {
-			xbps_set_cb_state(xhp, XBPS_STATE_CONFIG_FILE,
-			    0, pkgver,
-			    "Updating configuration file `%s' provided "
-			    "by `%s'.", cffile, pkgver);
+			if (!(xhp->flags & XBPS_FLAG_KEEP_CONFIG)) {
+				xbps_set_cb_state(xhp, XBPS_STATE_CONFIG_FILE,
+					0, pkgver,
+					"Updating configuration file `%s' provided "
+					"by `%s'.", cffile, pkgver);
+			} else {
+				version = xbps_pkg_version(pkgver);
+				assert(version);
+				snprintf(buf, sizeof(buf), ".%s.new-%s", cffile,
+					version);
+				xbps_set_cb_state(xhp, XBPS_STATE_CONFIG_FILE,
+					0, pkgver, "Preserving `%s',  installing configuration file to `%s'.",
+					cffile, buf);
+				archive_entry_copy_pathname(entry, buf);
+			}
```

Comment:
I think I'd rather this logic be moved to only have the higher level `else if`, where you "fall through" to basically this exact code farther down only if the new configuration file isn't changed.

Reply:
I'm not sure I understand what you mean. Do you want to add a check for the configuration option not being set to enter this if-else and then always enter the last if-else (with the almost identical code) unconditionally if the option is set?

Something along the linens of the snippet below?
```suggestion
else if ((!(xhp->flags & XBPS_FLAG_KEEP_CONFIG)) &&
         (strcmp(sha256_orig, sha256_cur) == 0) &&
         (strcmp(sha256_orig, sha256_new)) &&
	 (strcmp(sha256_cur, sha256_new))) {`		
```

Reply:
Something like that yes

Reply:
When considering how to write in the option, I did it the way it currently is because I liked the idea of the independent enumeration of configuration cases. I thought a minor code duplication was preferable over having the if-else-statements deal with more than one specific set of configuration clash as I don't like adding on "exceptions to the rules".

I'll see if I can write it in such a way that it doesn't get too convoluted, but I'm also open to the possibility that my preferences might be off.

Reply:
@Duncaen opinions?

Reply:
I contemplated putting the preservation functionality in its own function at first, but there are a lot of variables being used that would lead to an enormous argument list.

Depending on how you feel about context specific macros, one solution could be simply putting the function in a macro that takes no arguments (except for maybe the log message if we care to differentiate between the two cases) and counts on the scope having correctly named variables. That way code duplication would be reduced, while the current control flow is kept.

Reply:
Don't complicate it more than it is already, the suggested change above looks ok to me.

Reply:
Sure, done.